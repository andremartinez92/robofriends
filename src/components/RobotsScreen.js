// @flow
import React from 'react';

import Header from '../components/Header';
import SearchBox from '../components/SearchBox';
import RobotsList from '../components/RobotsList';

import { type RobotData } from '../types';

import './RobotsScreen.css';

type Props = {
    isPending: boolean,
    onSearchChange: (searchField: string) => void,
    robots: RobotData[],
}

function RobotsScreen({
    isPending,
    onSearchChange,
    robots
}: Props) {
    return (
        <div className="tc">
            <Header />
            <SearchBox onSearchChange={onSearchChange} />
            <RobotsList isPending={isPending} robots={robots} />
        </div>
    );
}

export default RobotsScreen;