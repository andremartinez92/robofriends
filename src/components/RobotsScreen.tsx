import React from 'react';

import Header from './Header';
import SearchBox from './SearchBox';
import RobotsList from './RobotsList';

import { RobotData } from '../types';

import './RobotsScreen.css';

interface Props {
    isPending: boolean,
    onSearchChange: (searchField: string) => void,
    robots: RobotData[],
}

function RobotsScreen({
    isPending,
    onSearchChange,
    robots
}: Props) {
    return (
        <div className="tc">
            <Header />
            <SearchBox onSearchChange={onSearchChange} />
            <RobotsList isPending={isPending} robots={robots} />
        </div>
    );
}

export default RobotsScreen;