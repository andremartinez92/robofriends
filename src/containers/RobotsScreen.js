// @flow

import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import './RobotsScreen.css';
import { type RobotData } from '../types';
import CardList from '../components/CardList';
import Header from '../components/Header';
import SearchBox from '../components/SearchBox';
import ScrollableSection from '../components/ScrollableSection';
import ErrorBoundary from '../components/ErrorBoundary';
import filterRobotsBySearch from '../helpers/filterRobotsBySearch';

import { changeSearchField } from '../middleware/actions';

function RobotsScreen() {
  const dispatch = useDispatch();
  const [robots, setRobots] = React.useState<RobotData[]>([]);

  const searchField: string = useSelector(state => state.searchField);

  const setSearchField: Function = (searchField: string) =>
    dispatch(changeSearchField(searchField));

  React.useEffect(() => {
    async function fetchData() {
      const response = await fetch(
        'https://jsonplaceholder.typicode.com/users'
      );
      const robots = await response.json();
      setRobots(robots);
    }

    fetchData();
  }, []);

  const filteredRobots: RobotData[] = filterRobotsBySearch(robots, searchField);

  return (
    <div className="tc">
      <Header />
      <SearchBox onSearchChange={setSearchField} />
      <ScrollableSection>
        <ErrorBoundary>
          <CardList robots={filteredRobots} />
        </ErrorBoundary>
      </ScrollableSection>
    </div>
  );
}

export default RobotsScreen;
