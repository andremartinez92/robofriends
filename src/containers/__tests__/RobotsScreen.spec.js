import React from 'react';
import { mount } from 'enzyme';
import ReactTestUtils from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import RobotsScreen from '../RobotsScreen';

const initialState = {
  search: { searchField: '' },
  robots: { isPending: false, robots: [] },
};

function createWrapper(store) {
  return mount(
    <Provider store={store}>
      <RobotsScreen />
    </Provider>
  );
}

// TODO
describe.skip('<RobotsScreen>', () => {
  let mockStore;
  beforeEach(() => {
    fetch.resetMocks();
    mockStore = configureMockStore([])(initialState);
  });

  describe('when mounted', () => {
    it('fetches the data from the expected URL', async () => {
      fetch.mockResponse(JSON.stringify([]));
      await ReactTestUtils.act(async () => {
        createWrapper(mockStore);
      });

      expect(fetch.mock.calls.length).toEqual(1);
      expect(fetch.mock.calls[0][0]).toEqual(
        'https://jsonplaceholder.typicode.com/users'
      );
    });
  });

  describe('when receives empty data from URL', () => {
    it('renders correctly', async () => {
      fetch.mockResponse(JSON.stringify([]));
      let wrapper;
      await ReactTestUtils.act(async () => {
        wrapper = createWrapper(mockStore);
      });

      const subject = wrapper.find('RobotsScreen');
      expect(subject).toMatchSnapshot();
      expect(subject.find('Card')).toHaveLength(0);
    });
  });

  describe('when receives data from URL', () => {
    const data = [
      { name: 'default', email: 'default@email1.com', id: 1 },
      { name: 'ninja', email: 'email@email2.com', id: 2 },
      { name: 'robot', email: 'email@email3.com', id: 3 },
    ];

    describe('when search field is empty', () => {
      it('renders all results', async () => {
        fetch.mockResponse(JSON.stringify(data));
        let wrapper;
        await ReactTestUtils.act(async () => {
          wrapper = createWrapper(mockStore);
        });

        wrapper.update();

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Card')).toHaveLength(3);
      });
    });

    describe('when search field has text', () => {
      it('renders only results that match search', async () => {
        mockStore = configureMockStore([])({ searchField: 'a' });

        fetch.mockResponse(JSON.stringify(data));
        let wrapper;
        await ReactTestUtils.act(async () => {
          wrapper = createWrapper(mockStore);
        });

        wrapper.update();

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Card')).toHaveLength(2);
      });
    });

    describe('when search field is changed', () => {
      it('dispatches the expected action', async () => {
        fetch.mockResponse(JSON.stringify(data));
        const spy = jest.spyOn(mockStore, 'dispatch');

        let wrapper;
        await ReactTestUtils.act(async () => {
          wrapper = createWrapper(mockStore);
        });

        const { onSearchChange } = wrapper.find('SearchBox').props();
        onSearchChange('a');

        expect(spy).toHaveBeenCalledWith({
          payload: { searchField: 'a' },
          type: 'CHANGE_SEARCH_FIELD',
        });
      });
    });
  });
});
