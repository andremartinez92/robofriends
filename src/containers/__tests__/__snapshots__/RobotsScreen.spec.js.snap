// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RobotsScreen> when receives data from URL when search field has text renders only results that match search 1`] = `
<div
  className="tc"
>
  <h1
    className="f1"
  >
    Robofriends
  </h1>
  <div
    className="pa2"
  >
    <input
      className="pa3 ba b--green bg-lightest-blue"
      onChange={[Function]}
      placeholder="Search robots"
      type="search"
    />
  </div>
  <div
    style={
      Object {
        "height": "800px",
        "overflow": "scroll",
      }
    }
  >
    Array [
      <div
        className="tc bg-light-green dib br3 pa3 ma2 grow bw2 shadow-5"
      >
        <img
          alt="robots"
          src="https://robohash.org/1?200x200"
        />
        <div>
          <h2>
            default
          </h2>
          <p>
            default@email1.com
          </p>
        </div>
      </div>,
      <div
        className="tc bg-light-green dib br3 pa3 ma2 grow bw2 shadow-5"
      >
        <img
          alt="robots"
          src="https://robohash.org/2?200x200"
        />
        <div>
          <h2>
            ninja
          </h2>
          <p>
            email@email2.com
          </p>
        </div>
      </div>,
    ]
  </div>
</div>
`;

exports[`<RobotsScreen> when receives data from URL when search field is empty renders all results 1`] = `
<div
  className="tc"
>
  <h1
    className="f1"
  >
    Robofriends
  </h1>
  <div
    className="pa2"
  >
    <input
      className="pa3 ba b--green bg-lightest-blue"
      onChange={[Function]}
      placeholder="Search robots"
      type="search"
    />
  </div>
  <div
    style={
      Object {
        "height": "800px",
        "overflow": "scroll",
      }
    }
  >
    Array [
      <div
        className="tc bg-light-green dib br3 pa3 ma2 grow bw2 shadow-5"
      >
        <img
          alt="robots"
          src="https://robohash.org/1?200x200"
        />
        <div>
          <h2>
            default
          </h2>
          <p>
            default@email1.com
          </p>
        </div>
      </div>,
      <div
        className="tc bg-light-green dib br3 pa3 ma2 grow bw2 shadow-5"
      >
        <img
          alt="robots"
          src="https://robohash.org/2?200x200"
        />
        <div>
          <h2>
            ninja
          </h2>
          <p>
            email@email2.com
          </p>
        </div>
      </div>,
      <div
        className="tc bg-light-green dib br3 pa3 ma2 grow bw2 shadow-5"
      >
        <img
          alt="robots"
          src="https://robohash.org/3?200x200"
        />
        <div>
          <h2>
            robot
          </h2>
          <p>
            email@email3.com
          </p>
        </div>
      </div>,
    ]
  </div>
</div>
`;

exports[`<RobotsScreen> when receives empty data from URL renders correctly 1`] = `
<div
  className="tc"
>
  <h1
    className="f1"
  >
    Robofriends
  </h1>
  <div
    className="pa2"
  >
    <input
      className="pa3 ba b--green bg-lightest-blue"
      onChange={[Function]}
      placeholder="Search robots"
      type="search"
    />
  </div>
  <div
    style={
      Object {
        "height": "800px",
        "overflow": "scroll",
      }
    }
  />
</div>
`;
